### ISSUE TITLE
Feature request: FilterCode for (not) completing activities

### ISSUE THREAD

**Issue by mattgig** [michael-milette/moodle-filter_filtercodes#152]
At the moment there is a filter code which allows content to be displayed when you belong to a certain group: {ifingroup id}.

It would be great, if there were two new filter codes which displays content when an activity has been completed or not: {ifactivitycompleted id} / {ifacitivitynotcompleted id}

With such filter codes you could for instance provide a revision for questions in a test which have not been solved properly, without having to fill in every question into a separate label or page. Or you could adjust the content of a book or other resource on the level of a sentence or paragraph according to the mastery a student has shown in the past or the choices he/she had made earlier on.

I don't really know whether something like this is doable, without running into problems or having to write large junks of code, but it would certainly allow teachers to personalise the learning experience of students to a very high degree without having to rewrite whole resources as a whole.

**Comment by michael-milette** [comment]
Hi @mattgig ,

Thanks for sharing your ideas. I would need to better understand your vision to ensure that such a tag would be feasible. Tags are often context sensitive and may not work as expected or at all in some situations.

Imagine that the {ifactivitycompleted id} tag was a reality. Could you walk me through the step-by-step process of setting this up how it would be used? If you have multiple scenarios in in mind, please describe them (e.g. used in different types of activities). 

Best regards,

Michael

**Comment by mattgig** [comment]
Hi Michael

Thanks for showing interest in a rather special feature request.

A filter like that could be useful for a course scenario I'm writing at the moment. In a programming course about microcontrollers students use different hardware. As I want to streamline instructions, I only want them so see instructions which are relevant to the 
hardware they are using. On the other hand, I don't want to copy the same material which is suitable for all students into different pages as this makes adjustment tedious to manage.
At the moment, there are two ways to achieve this. You either adjust course content by the level of a whole activity or you make use of groups (we use the plugin group choice so that we don't have to manage the groups) and than you can adjust the content of any page using your plugin in combination with the {ifingroup} filter.

If a {ifactivitycompleted} filter existed, you could achieve the same thing depending on activity completions. This would allow for very fine-grained adjustments of resources presented to students.

The {ifactivitynotcompleted} filter could be used for a scenario I'm not able to achieve at the moment: After a test, students get a chance to rehearse all the questions they haven't solved properly yet by including them into a page with the embedding question filter. Because there is also a plugin which makes use of single question completion, you could put all the question into that page, but each student would only see the question he/she still needs to practice. I think this could be a very nice feature in the long run.

So basically, I want to use the filter like a language filter, but based on past activities and depending on the scenario down to individual words, not whole activities.

I don't know whether the scenarios I've described makes sense to you and I certainly don't know whether something like this is within a managable scope of the plugin. In case you need more information about it or an example course, I could provide you with a demo course making use of the {ifingroup} filter.

**Comment by mattgig** [comment]
Hi Michael

Thanks a lot for implementing this feature. 

And I just want to let you know that in the German forum somebody now uses the feature to generate detailed certificates based on the activities that person has done.

See: https://moodle.org/mod/forum/discuss.php?d=435828

That's yet another use case for a great plugin.
